design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,506 @@
+-- fails as key column not in projection.
+
+CREATE TABLE OUTPUT AS SELECT COUNT(*) FROM INPUT GROUP BY V1;
+-- fails as key column not in projection.
+
+CREATE STREAM OUTPUT AS SELECT V0, ID FROM INPUT PARTITION BY V1;
+-- fails as key column not in projection.
+
+CREATE TABLE OUTPUT AS SELECT I1.V0, I2.V2 FROM INPUT I1 JOIN INPUT I2 ON I1.ID = I2.ID;
+-- fails as key column not in projection.
+```
+
+We propose this is acceptable, and preferable to the current model.
+
+There is no reason why ksqlDB should not support creating sources in the future that do not contain
+the key column. See the [rejected alternatives](#rejected_alternatives) section for more info.
+
+### Allowing the key column to be copied to the value
+
+Hopefully, all of the above seems simple and clear. Now for one of the more tricky and murky bits...
+
+The observant among you may be thinking 'ah, but what if I _want_ to put a copy of the key column
+into the value?'. With current syntax you can do:
+
+```sql
+-- current syntax:
+CREATE TABLE OUTPUT AS SELECT V0, V1, ID AS V2 FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
+```
+
+With the proposed syntax the `ID AS V2` would be treated an aliased key column, resulting in the
+schema 'V2 INT KEY, V0 INT, V1 INT'. So, how does the user create an copy of the key in the value,
+if that's what they need to do?
+
+We propose introducing a `AS_VALUE` function that can be used to
+indicate the key column should be copied as a value column. For example,
+
+```sql
+CREATE TABLE OUTPUT AS SELECT ID, V0, V1, AS_VALUE(ID) AS V2 FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
+```
+
+### Grouping by multiple expressions.
+
+KsqlDB supports grouping by multiple expressions, for example:
+
+```sql
+SELECT V0, ABS(V1), COUNT(*) AS COUNT FROM INPUT GROUP BY V0, ABS(V1);
+```
+
+However, it does not _yet_ support multiple key columns. If the above is converted to a persistent
+query the key is generated by concatenating the string representation of the grouping expressions.
+For example:
+
+```sql
+CREATE TABLE OUTPUT AS SELECT V0, ABS(V1) AS V1, COUNT(*) AS COUNT FROM INPUT GROUP BY V0, ABS(V1);
+-- resulting schema: KSQL_COL_0 STRING KEY, V0 INT, V1 INT, COUNT BIGINT
+-- where KSQL_COL_0 contains data in the form V0 + "|+|" + ABS(V1)
+```
+
+Even though ksqlDB is currently combining the multiple grouping expressions, we propose that the
+projection should still accept the individual columns, and recognise them as key columns. This will
+be compatible with the upcoming multiple-key-column support.
+
+However, this posses a problem, as it does not provide a single place where the user can provide an
+alias for the system generated `KSQL_COL_0` key column name. Any solution to allow providing an
+alias would likely be incompatible with the planned multiple key column support. 
+
+Hence, we propose leaving this edge case unsolved, i.e. users will _not_ be able to provide an alias
```

Comment:
As the proposal is to use `ROWKEY` column for outer joins, it seems the same pattern could be applied for this case to allow people to rename the PK?

Reply:
Hummmm.... you've got a point.  

However, I think there's a subtle difference we may want to consider.

With a grouping statement on multiple things, the current implementation combines all the result columns into a single `STRING` value.  So the Kafka message's key contains all the grouping data, just in a nasty format.  No additional column is being synthesised.

Conversely, for outer joins the Kafka message's key contains the result of `COALESCE(leftJoinExp, rightJoinExp)`, i.e. unlike the grouping statement, it does _NOT_ contain all the joining data: it loses the data about any side being `null`.  

Because of this subtle difference we know that the upcoming structured keys work will allow users to alias the multiple grouping expressions in the projection. However, the issue with joins can not be fixed with structured key support alone. 

If we go the same route for groupings as we've proposed for joins, then we end up with:

```sql
CREATE TABLE OUTPUT AS
   SELECT ROWKEY AS K, COUNT() FROM INPUT GROUP BY V0, V1;

-- vs --

CREATE TABLE OUTPUT AS
   SELECT V0, V1, COUNT() FROM INPUT GROUP BY V0, V1;
```

There's certainly arguments for going either way. Personally, I'm happy with what's been proposed for groupings in the KLIP. 

cc @derekjn for a product view.

Reply:
I guess we agree that the end-state should be the last query you showed:
```sql
CREATE TABLE OUTPUT AS
   SELECT V0, V1, COUNT() FROM INPUT GROUP BY V0, V1;
```
(With a proper structured key `<V0,V1>` stored in the message key).

I guess the question is what intermediate state we want to be in. The query from above is a valid query _now_ however, it does not really expose the PK that is stored in the `ROWKEY`.

From the KLIP:
> we propose that the projection should still accept the individual columns, and recognise them as key columns

Not sure if I can follow. If both columns V0 and V1 are store in the value, how can this be done?

As this KLIP seems to try to expose the actual message key in the schema, it seems consequent to add `ROWKEY` for this case as an intermediate step. I understand, that it might look like a step backward for a language POV as the above query that is valid now, would not be valid until we reach the end-state and it becomes valid again...

Reply:
To be clear the KLIP is proposing that the following will be valid:

```sql
CREATE TABLE OUTPUT AS
   SELECT V0, V1, COUNT() FROM INPUT GROUP BY V0, V1;
```

And both `V0` and `V1` will be stored in the key in a munged together column called `KSQL_COL_0`. 

Yes, we could support:

```sql
CREATE TABLE OUTPUT AS
   SELECT KSQL_COL_0, COUNT() FROM INPUT GROUP BY V0, V1;
```

But sures will be left wondering where `KSQL_COL_0` came from.

Given than we're adding support for multiple key columns next, product are happy with not supporting aliasing of the resulting key column name.